Don't forget the datatype when decoding a PEM file

mattcaswell opened the pull request:
The OSSL_STORE code was forgetting the datatype that we read from the
PEM header when decoding the DER.

Fixes #13046

WIP because it needs a test.

@levitte - I'm not sure where to put a test for this? Can you give me a hint?

levitte commented:
> @levitte - I'm not sure where to put a test for this? Can you give me a hint?

I would write a separate recipe for this, that generates a X9.42 and a DSA PEM file, and capture the output of `openssl storeutl` on them, and see that you get the BEGIN line you expect for each.

mattcaswell commented:
I've rebased this to fix a conflict with master and extended it with additional tests.

My first attempt at adding tests was to create a test that called various PEM_read_bio_* functions to check that we got the expected result back. This effectively does the same thing as the reproducer in #13046. To my surprise I found that the tests I had written were passing in master even without my fix applied!! On further investigation I found that issue in #13046 no longer exists. It was "fixed" by commit 35426b2f9c. However the "fix" only addresses the symptoms by changing `PEM_read_bio_DHparams()` to not use OSSL_STORE anymore. The bug in OSSL_STORE remains.

Therefore I have kept the original tests that I wrote (they would have caught the issue before 35426b2f9c was applied). I've also added more tests to test OSSL_STORE directly.

This PR is now out of WIP. Please take a look.

mattcaswell commented:
Fixup pushed to address some missing no-dh and no-dsa guards.

levitte commented on test/pem_read_depr_test.c:
```diff
@@ -0,0 +1,213 @@
+
+#include "testutil.h"
+
+const char *datadir;
```
Add `static`

mattcaswell replied:
Done

levitte commented:
I saw a surprising memleak in the last Travis results...  have you tried an ASAN build?

mattcaswell commented:
> I saw a surprising memleak in the last Travis results... have you tried an ASAN build?

Leak was in the test. Now fixed.

levitte commented on test/ossl_store_test.c:
```diff
@@ -76,9 +77,10 @@ static int get_params(const char *uri, const char *type)
+    EVP_PKEY_free(pkey);
 
-    return 0;
+    return ret;
```
The `err` section doesn't look very "errory" any more.  How about changing all from `err:` to this?

``` C
    if (pkey != NULL)
        ret = EVP_PKEY_is_a(pkey, type);
    EVP_PKEY_free(pkey);

 err:
    OSSL_STORE_close(ctx);
    return ret;
```

mattcaswell replied:
Done

levitte reviewed (approved):
If CIs agree...

mattcaswell commented:
Travis failures are unrelated to this PR.

levitte commented:
> Travis failures are unrelated to this PR.

Confirmed

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

slontis commented:
@mattcaswell are you saying that this PR doesnt fix the ossl store issue? If so is there an issue for this?

mattcaswell commented:
>  @mattcaswell are you saying that this PR doesnt fix the ossl store issue? If so is there an issue for this?

I assume you are referring to this comment by me:

>My first attempt at adding tests was to create a test that called various PEM_read_bio_* functions to check that we got the expected result back. This effectively does the same thing as the reproducer in #13046. To my surprise I found that the tests I had written were passing in master even without my fix applied!! On further investigation I found that issue in #13046 no longer exists. It was "fixed" by commit 35426b2. However the "fix" only addresses the symptoms by changing PEM_read_bio_DHparams() to not use OSSL_STORE anymore. The bug in OSSL_STORE remains.

What I was trying to say (apparently unsuccessfully) is that there has been a "fix" already applied that meant the reproducer in #13046 no longer works. But that other "fix" did not address the OSSL_STORE bug. *This* PR *does* address that bug.

mattcaswell commented:
Pushed. Thanks.
